streamingccl: initial async flushing on the consumer

stevendanna opened the pull request:
Here, we allow the stream ingestion processor to flush one batch while building the next batch.

In an ad-hoc single-node to single-node test, this doubles throughput on TPCC. However, we still see substantial throughput falloff after the first minute or so.

To facilitate async flushing, the stream ingestion processor now reads data from the merged subscription in a goroutine rather than directly in Next(). Next() returns the last result from the most recently completed flush.

All errors are now returned to Next() via an err channel, with the first error winning.

Release note: None

Epic: none

stevendanna committed:
streamingccl: initial async flushing on the consumer

Here, we allow the stream ingestion processor to flush one batch while
building the next batch.

In an ad-hoc single-node to single-node test, this doubles throughput
on TPCC. However, we still see substantial throughput falloff after
the first minute or so.

To facilitate async flushing, the stream ingestion processor now reads
data from the merged subscription in a goroutine rather than directly
in Next(). Next() returns the last result from the most recently
completed flush.

All errors are now returned to Next() via an err channel, with the
first error winning.

Release note: None

Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104326)
<!-- Reviewable:end -->

stevendanna commented:
@lidorcarmel This is slightly WIP, but I wanted to get it out for review so I can hopefully get it in soon.

stevendanna committed:
streamingccl: fix flaky test

It's unclear why this was working before; but this assertion assumes
that the merged stream will return message from each stream in a
"fair" order.  But there is nothing that prevents all of partition 1
messages being processed first. As a result, we may never see a
frontier with key_1 at ts4 since it will be at ts6 on the very first
flush.

The comment seems to imply that this was the intent originally.

Epic: none

Release note: none

lidorcarmel commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -385,42 +456,33 @@ func (sip *streamIngestionProcessor) Next() (rowenc.EncDatumRow, *execinfrapb.Pr
+			return row, nil
+		}
+	case err := <-sip.errCh:
 		if err != nil {
```
I think err cannot be nil here, right? if it can then we will drop the second error which is not great..

stevendanna replied:
It cannot. I removed this and moved an `if err == nil` into sendError just as a bit of precaution.

lidorcarmel commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -438,107 +500,102 @@ func (sip *streamIngestionProcessor) close() {
+	if sip.mergedSubscription != nil {
+		sip.mergedSubscription.Close()
+	}
+	if sip.stopCh != nil {
```
can it be nil?

lidorcarmel commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -438,107 +500,102 @@ func (sip *streamIngestionProcessor) close() {
+		close(sip.stopCh)
+	}
+
+	// We shouldn't need to explicitly cancel the context for
+	// members for the worker group. The mergedSubscription close
```
nit: maybe 'for members of'?

lidorcarmel reviewed (approved):
really nice and important pr. it might be a bit too much for the release tomorrow though.

I only had a few nits..

re testing - can you add a simple test file for the new merged_subscription.go file? the test I have in mind would be a bit silly, hopefully you will have better ideas.. but anything would be fine. this way we will be encouraged to add tests when we make that file more complex.

last - do you think that our unit tests for stream_ingestion_processor.go are good enough? and cover the new code well? just wondering, not really suggesting more tests..

stevendanna commented:
> re testing - can you add a simple test file for the new merged_subscription.go file? the test I have in mind would be a bit silly, hopefully you will have better ideas.. but anything would be fine. this way we will be encouraged to add tests when we make that file more complex.

Done.

> last - do you think that our unit tests for stream_ingestion_processor.go are good enough? and cover the new code well? just wondering, not really suggesting more tests..

Thanks for the nudge. I think we need to improve them.  I took a look at the coverage of the go-level tests we have. 

<img width="1875" alt="Screenshot 2023-06-06 at 14 37 01" src="https://github.com/cockroachdb/cockroach/assets/852371/cfab1597-28d5-4c59-986c-e509cf42fa8f">

Most of the missing coverage are error conditions. I added a few test cases to improve the stream_ingestion_processor a few percent, and in the process discovered a bug, which is fixed in the last commit.

I'll rebase the commits before merging.

lidorcarmel reviewed (approved):
looks like there a test failure, but other than that lgtm.

stevendanna commented:
bors r=lidorcarmel

stevendanna commented:
bors retry

stevendanna commented:
bors retry
